pkg/cmd/validate.go

```diff
@@ -164,39 +189,27 @@ func (c *commandFlags) Run(cmd *cobra.Command, args []string) error {
 		return err
 	}
 
+	filesErrors := make(FilesErrors)
+	for _, path := range files {
+		errors := ValidateFile(path, factory)
+		filesErrors[path] = errors
+	}
+
 	if c.outputFormat == OutputHuman {
-		for _, path := range files {
-			fmt.Fprintf(cmd.OutOrStdout(), "\n\033[1m%v\033[0m...", path)
-			var errs []error
-			for _, err := range ValidateFile(path, factory) {
-				if err != nil {
-					errs = append(errs, err)
-				}
-			}
-			if len(errs) != 0 {
-				fmt.Fprintln(cmd.OutOrStdout(), "\033[31mERROR\033[0m")
-				for _, err := range errs {
-					fmt.Fprintln(cmd.ErrOrStderr(), err.Error())
-				}
-			} else {
-				fmt.Fprintln(cmd.OutOrStdout(), "\033[32mOK\033[0m")
-			}
+		if err := printHumanErrors(filesErrors, cmd.OutOrStdout(), cmd.ErrOrStderr()); err != nil {
+			return err
 		}
 	} else {
-		res := map[string][]metav1.Status{}
-		for _, path := range files {
-			for _, err := range ValidateFile(path, factory) {
-				res[path] = append(res[path], errorToStatus(err))
-			}
-		}
-		data, e := json.MarshalIndent(res, "", "    ")
-		if e != nil {
-			return fmt.Errorf("failed to render results into JSON: %w", e)
+		if err := printJsonErrors(filesErrors, cmd.OutOrStdout(), cmd.ErrOrStderr()); err != nil {
+			return err
 		}
-		fmt.Fprintln(cmd.OutOrStdout(), string(data))
 	}
 
-	return nil
+	if filesErrors.hasError() {
+		return errors.New("found some errors in the manifests")
+	} else {
+		return nil
+	}
```

Comment:
I'm not super excited about that.

Something you could do is:
1. Give FilesErrors a  "Append" function that supports _self_ being nil
3. return the instance here without checking for nil, nor creating a dummy error

So probably something like this:
```golang
type Error struct {
    path string
    err error
}

type FilesErrors []Error

func (e *FilesError) Append(path string, err error) *FilesError {
    if err == nil {
        return e
    } 
    return append(e, Error{path: path, err: err})
}

func (e *FilesError) Error(format string) string {
    // print according to format, can call separate functions if needed.
}

... // In Run()
var errs *FileErrors // Get a nil pointer to errors
for _, path := range files {
    errs = errs.Append(path, ValidateFile(path, factory))
}
...
return errs // Will be null if we've never added any error.

... // in main()
if err := rootCmd.Execute(); err != nil {
    fmt.Fprintln(os.Stderr, err.Error(rootCmd.outputFormat))
    os.Exit(1)
}
```
I think this has a lot less conditions to keep track of.

Reply:
I generally agree, but there are some considerations. I start with two of them which are the most important ones:

First, each file can contain multiple documents.[^1] `ValidateFile` function generates an error (possibly `nil`) for each document and returns a list of errors. [^2] So, it's important to keep track of `nil` errors as json formatter needs it.

Second, in order for the `FilesErrors` type to be considered a golang error, we have to implement `Error` method without any arguments. So, what's your opinion about declaring `FilesErrors` as a `struct` with `errors` and `outputFormat` fields, and using `outputFormat` field to implement the logic of `Error()` method?

[^1]: It runs `ValidateDocument` over all documents of the file, which returns exactly one error per document. So, `ValidateFile` returns a slice containing those errors, i'th error for the i'th document.
[^2]: I think the human output is a little ambiguous when a file contains multiple documents. For example, if the first and third documents in a file have an error and the second document is sane, the human formatter prints the first and third errors and prints nothing about the sanity of the second document. However, this issue is out of scope and I intended to address it in another issue/pr.

Reply:
Sorry for the delay, code freeze in k/k is approaching and I lost track of that.

Yeah, I made the invalid assumption that we weren't printing anything if there was no error.
Considering that, I don't see how we can return a relevant error out of the `Run` method, maybe it's fine to just `os.Exit` directly from it, and always return nil when no errors occur?

Reply:
I'd prefer returning an `errors.Join` a slice of the `Error` type @apelisse posted above, then in main we can exit if the execution returns non-nil error.